### PR TITLE
refactor(core): service discovery mechanism

### PR DESCRIPTION
**Pull request by kkopanidis**
This PR attempts to fix SD inconsistencies, with modules being advertised as unhealthy or not even registered when they are online. It might need a couple more changes to make sure, but initial testing seems to show that the issues are resolved.

Currently includes several logs to assist with testing. Should remove before merging.

**What kind of change does this PR introduce?** 

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [x] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [ ] Yes
- [x] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `main` branch
- [x] When resolving a specific issue, it's referenced in the PR's description  (e.g. `fix #xxx`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature 

**Other information:**

### CODE REVIEW

File: libraries/grpc-sdk/src/utilities/StateManager.ts [hunk 1]
```diff
@@ -1,11 +1,56 @@
...
+      // to improve performance under high contention
+      // see https://www.awsarchitectureblog.com/2015/03/backoff.html
+      retryJitter: 200, // time in ms
+      // The minimum remaining time on a lock before an extension is automatically
+      // attempted with the `using` API.
+      automaticExtensionThreshold: 500, // time in ms
+    });
+  }
+
+  async acquireLock(resource: string, ttl: number = 5000): Promise<Lock> {
```
Comment (kon14) [inline libraries/grpc-sdk/src/utilities/StateManager.ts:36]: `resource` type should probably be `KNOWN_LOCKS`

File: packages/core/src/config-manager/service-discovery/ServiceMonitor.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    delete this.moduleHealth[moduleName];
+    this._serviceRegistry.removeModule(moduleName);
+    this.reviveService(moduleName, moduleUrl);
+  }
+
+  updateModuleHealth(
+    moduleName: string,
+    moduleUrl: string,
+    moduleStatus: HealthCheckStatus,
+    broadcast = true,
```
Comment (kon14) [inline packages/core/src/config-manager/service-discovery/ServiceMonitor.ts:99]: unused arg

**Review by kon14 — CHANGES_REQUESTED** [review]
Upon initial testing, it seems to work as expected.

Restarting `Router`/`Database` didn't cause any unrecoverable issues.
However, upon successively restarting both of the above a few times I did notice some race-condition. The deployment still recovered, and routes seemed to be functional after a short while though.
Database output:

```
[2023-11-23 19:08:55] [ERROR]: Failed to register routes for module
[2023-11-23 19:08:55] [ERROR]: /router.Router/RegisterConduitRoute INTERNAL: Well that failed :/ ClientError: /router.Router/RegisterConduitRoute INTERNAL: Well that failed :/
```

File: libraries/grpc-sdk/src/utilities/StateManager.ts (hunk 1 above)
Comment (kkopanidis) [reply libraries/grpc-sdk/src/utilities/StateManager.ts:36]: No, known locks is used for ease, you can acquire lock in whatever you want

**Review by kon14 — APPROVED** [review]

**Pull request merged**